src/System.Private.CoreLib/src/System/Threading/ClrThreadPool.WaitThread.cs

```diff
@@ -0,0 +1,425 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Diagnostics;
+using Internal.Runtime.Augments;
+using System.Runtime.InteropServices;
+
+namespace System.Threading
+{
+    internal partial class ClrThreadPool
+    {
+        /// <summary>
+        /// A linked list of <see cref="WaitThread"/>s.
+        /// </summary>
+        private WaitThreadNode _waitThreadsHead;
+        private WaitThreadNode _waitThreadsTail;
+
+        private LowLevelLock _waitThreadListLock = new LowLevelLock();
+
+        /// <summary>
+        /// Register a wait handle on a <see cref="WaitThread"/>.
+        /// </summary>
+        /// <param name="handle">A description of the requested registration.</param>
+        internal void RegisterWaitHandle(RegisteredWait handle)
+        {
+            bool needToStartWaitThread = RegisterWaitHandleOnWaitThread(handle);
+
+            if (needToStartWaitThread)
+            {
+                _waitThreadsTail.Thread.Start(); // If this fails, propogate the exception up like CoreCLR
```

Comment:
In case of any exception, the wait thread needs to be removed from the list. By that time, another thread may have registered another wait to the same thread before it was started.